### PR TITLE
Behaviors: Extend Global Styles API to read/write behaviors config.

### PR DESCRIPTION
**Pull request by cbravobernal**
## What?

Include a Behaviors selector in the Site Editor.

## Why?
Without this PR, the only way to set behaviors globally for all images was by editing `theme.json`. This PR enables a selector on the site editor, under `styles -> blocks -> image -> advanced` that uses Global Styles API to apply the behavior to all the images in the site.

### To-do list
- [x] Add behaviors top level to the REST API
- [x] Add the selector, set it as experimental
- [x]  Save into DB the behaviors of the site editor
- [x] Add animation options
- [x] Handle UI visibility with settings in theme.json
- [x] Get the animation value for the theme in the selector
- [x] Combine behaviors and styles in apply globally.
- [x] Apply globally does not work if default option is chosen as behavior.
- [x] Check if `undo` option is working correctly.
- [x] Make it work on revisions.
- [ ] Add e2e tests -> Moved to #52809
- [x] Code cleaning and refactor

### CODE REVIEW

**Comment by SantosGuillamot** [comment]
I've been taking a quick look at it and it looks great so far 👏 I have a couple of comments:

* It seems it is working fine when editing the behaviors in the block settings but it doesn't seem to be working for me in the global styles of the Image. I made a quick video to explain it:

https://github.com/WordPress/gutenberg/assets/34552881/126496b3-f291-4205-b2ef-0ca2d984285c

* Maybe we should make it work with the "Apply globally" button? I mean this one:

Screenshot 2023-07-10 at 10 54 09

**Comment by cbravobernal** [comment]
> It seems it is working fine when editing the behaviors in the block settings but it doesn't seem to be working for me in the global styles of the Image. I made a quick video to explain it.

That's weird, as I've been developing it with the case it doesn't work on your side. Not the first one 😅 
Will take a look at that!

**Comment by cbravobernal** [comment]
I think the error is that if you set it up on block settings, it doesn't work on styles after and viceversa.

**Comment by SantosGuillamot** [comment]
Btw, this is the error I get in the console, in case it is helpful:

Screenshot 2023-07-10 at 11 48 19

**Comment by SantosGuillamot** [comment]
> I think the error is that if you set it up on block settings, it doesn't work on styles after and viceversa.

I have the default option in the block settings and still doesn't work in the global styles 🤔

**Comment by cbravobernal** [comment]
I just added 3 tasks more regarding the `Apply Globally` option.

**Comment by cbravobernal** [comment]
All tests are 🟢 . Time to add some more and refactor. Feel free to comment on them!

**Comment by michalczaplinski** [comment]
This seems to test well for me! I've tested by creating a post with 4 images:
- With "default" behavior
- An image with "zoom" animation
- An image with "fade" animation
- An image with "no behaviors"

1. Changing the behavior in the Global Styles only changes it for the image with "default" behavior. 👍 
2. Updating the value for behaviors in the `theme.json`, also only changes the behavior for the image with "default" behavior 👍 .
3. Adding an image in the site editor, adding a behavior to it and clicking "Apply Globally" works fine and applies the behaviors to all blocks 👍 
4. ❌ The one bit that does not seem to work 100% are the style revisions. Whenever, I try to go back to a previous revision, the behaviors are reset to "default" rather than go to the specific value defined in that revision. I've recorded a video with the issue:

https://github.com/WordPress/gutenberg/assets/5417266/4021d4e4-546f-4e0a-afad-6b7c97c53a5a

I've pushed a few small commits to clean up the code a little bit, and would be great to include a few e2e tests for this as well (but you already mentioned this). 

But in general, awesome work, @c4rl0sbr4v0  👏 🙂

File: packages/block-editor/src/components/global-styles/use-global-styles-output.js [hunk 1]
```diff
@@ -1217,7 +1217,7 @@ export function useGlobalStylesOutputWithConfig( mergedConfig = {} ) {
 				} );
 			}
 		} );
-
+		// we need to introduce behaviors here
```
Comment (cbravobernal) [inline packages/block-editor/src/components/global-styles/use-global-styles-output.js]: TODO 😄

**Comment by michalczaplinski** [comment]
I don't quite understand what's happening with this test failure: 

```
There was 1 failure:

1) Gutenberg_REST_Global_Styles_Revisions_Controller_Test::test_get_item_schema
Schema properties array has exactly 9 elements.
Failed asserting that actual size 10 matches expected size 9.
```

I was going to change [this line](https://github.com/WordPress/gutenberg/blob/ebe60b44c0323ee052ee716007dd85cc97d2a650/phpunit/class-gutenberg-rest-global-styles-revisions-controller-test.php#L361-L362) from `9` to `10` but I see that you've already done it before so there must be some reason for it 🙂 .

File: packages/block-editor/src/components/global-styles/utils.js [hunk 2]
```diff
@@ -415,6 +415,7 @@ export function areGlobalStyleConfigsEqual( original, variation ) {
 	}
 	return (
 		fastDeepEqual( original?.styles, variation?.styles ) &&
-		fastDeepEqual( original?.settings, variation?.settings )
+		fastDeepEqual( original?.settings, variation?.settings ) &&
+		fastDeepEqual( original?.behaviors, variation?.behaviors )
```
Comment (michalczaplinski) [inline packages/block-editor/src/components/global-styles/utils.js:419]: This is making [the e2e test](https://github.com/WordPress/gutenberg/blob/ebe60b44c0323ee052ee716007dd85cc97d2a650/test/e2e/specs/site-editor/style-variations.spec.js#L55-L58) fail because `original.behaviors === {}` and `variation.behaviors === undefined`. 

Are we sure that this is needed? If yes, then the fixtures for the test probably need to be updated.

**Comment by cbravobernal** [comment]
> 

Will do, once is working 😆 
I love this automation!

File: packages/block-editor/src/components/global-styles/utils.js (hunk 2 above)
Comment (cbravobernal) [reply packages/block-editor/src/components/global-styles/utils.js:419]: We need it, otherwise revisions with only behaviors won't work.

**Comment by michalczaplinski** [comment]
@ramonjd Could I ask you to take a look at the unit test failures in this PR? @c4rl0sbr4v0 and I have been trying to figure it out for a couple hours now but no dice. I'm asking you because you worked on Global Style Revisions and this is a related unit test.

- [This](https://github.com/WordPress/gutenberg/actions/runs/5588193085/jobs/10214599005?pr=52370#step:14:240) is the failure I'm talking about which corresponds to [this test](https://github.com/WordPress/gutenberg/blob/e0023fc32ace7659e90f7cec767077f6c41bc7e6/phpunit/class-gutenberg-rest-global-styles-revisions-controller-test.php#L355-L356).
- It seems to fail on the CI on WP 6.1 in either of those two tests:
   Screenshot 2023-07-18 at 15 18 38
- Changing the assertion on [this line](https://github.com/WordPress/gutenberg/blob/e0023fc32ace7659e90f7cec767077f6c41bc7e6/phpunit/class-gutenberg-rest-global-styles-revisions-controller-test.php#L361-L362) from `9` to `10` does not actually fix the problem (despite what I had expected). Instead, it causes another [unit test failure](https://github.com/WordPress/gutenberg/actions/runs/5588595403/jobs/10215627841?pr=52370#step:14:177) where the reverse is expected (actual size = 9,  expected size = 10)

**Comment by michalczaplinski** [comment]
Despite the unit test failures, the revisions are testing fine for me now 👍

File: packages/block-editor/src/components/global-styles/utils.js (hunk 2 above)
Comment (michalczaplinski) [reply packages/block-editor/src/components/global-styles/utils.js:419]: gotcha. I found the issue, we needed to add a behaviors property here: [ea1d540](https://github.com/WordPress/gutenberg/pull/52370/commits/ea1d54053ce4941a6049cef437c4047ba17fbaa7)

File: lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php [hunk 3]
```diff
@@ -282,14 +282,17 @@ public function prepare_item_for_response( $post, $request ) {
 		$fields = $this->get_fields_for_response( $request );
 		$data   = array();
 
-		if ( ! empty( $global_styles_config['styles'] ) || ! empty( $global_styles_config['settings'] ) ) {
+		if ( ! empty( $global_styles_config['styles'] ) || ! empty( $global_styles_config['settings'] ) || ! empty( $global_styles_config['behaviors'] ) ) {
```
Comment (ramonjd) [inline lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php]: So because the changes in this PR won't be in Wordpres 6.3, we'll need to create a new compat folder and move the changes into there, e.g., lib/compat/wordpress-6.4/class-gutenberg-rest-global-styles-revisions-controller-6-4.php

Then we can extend 6-3, e.g., `class Gutenberg_REST_Global_Styles_Revisions_Controller_6_4 extends Gutenberg_REST_Global_Styles_Revisions_Controller_6_3 {`, making sure to move the corresponding hook that registers routes via `rest_api_init` to lib/compat/wordpress-6.4/rest-api.php

File: lib/compat/wordpress-6.3/rest-api.php [hunk 4]
```diff
@@ -65,7 +65,7 @@ function gutenberg_register_global_styles_revisions_endpoints() {
  * Registers the Global Styles REST API routes.
  */
 function gutenberg_register_global_styles_endpoints() {
-	$global_styles_controller = new Gutenberg_REST_Global_Styles_Controller_6_3();
+	$global_styles_controller = new Gutenberg_REST_Global_Styles_Controller();
```
Comment (ramonjd) [inline lib/compat/wordpress-6.3/rest-api.php]: `gutenberg_register_global_styles_endpoints` should be moved to lib/experimental/rest-api.php

**Review by ramonjd — COMMENTED** [review]
> Could I ask you to take a look at the unit test failures in this PR? @c4rl0sbr4v0 and I have been trying to figure it out for a couple hours now but no dice. I'm asking you because you worked on Global Style Revisions and this is a related unit test.

Thanks for the ping.

For some reason it seems `get_item_schema()` from Core is being called first. I ran out of time to debug, but I'll carry on tomorrow. 

There must be a way around it, I'm sure we've done it before 🤔 

Also, in relation to extending the revisions endpoint controller, there are a couple of `private` methods and vars in there. I need to change the to `protected`. I'll throw up a PR to do that.

**Comment by ramonjd** [comment]
> Also, in relation to extending the revisions endpoint controller, there are a couple of private methods and vars in there. I need to change the to protected. I'll throw up a PR to do that.

Oh, if this doesn't make it into WordPress 6.3, then we'd have to copy over any private methods to new 6-4 classes. Not a big deal, but a bit of a pain. 

- https://github.com/WordPress/gutenberg/pull/52748

File: lib/experimental/class-gutenberg-rest-global-styles-controller.php [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+		if ( rest_is_field_included( 'settings', $fields ) ) {
+			$data['settings'] = ! empty( $config['settings'] ) && $is_global_styles_user_theme_json ? $config['settings'] : new stdClass();
+		}
+
+		if ( rest_is_field_included( 'styles', $fields ) ) {
+			$data['styles'] = ! empty( $config['styles'] ) && $is_global_styles_user_theme_json ? $config['styles'] : new stdClass();
+		}
+
+		if ( rest_is_field_included( 'behaviors', $fields ) ) {
```
Comment (ramonjd) [inline lib/experimental/class-gutenberg-rest-global-styles-controller.php]: I haven't tested this yet, but we might be able to get away with using `register_rest_field` instead. I'll have a closer look tomorrow.

File: lib/compat/wordpress-6.3/rest-api.php (hunk 4 above)
Comment (cbravobernal) [reply lib/compat/wordpress-6.3/rest-api.php]: I can do that quickly :-)

File: lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php (hunk 3 above)
Comment (cbravobernal) [reply lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php]: I did it in a non pushed commit, so I can push it.

File: lib/experimental/class-gutenberg-rest-global-styles-controller.php (hunk 5 above)
Comment (cbravobernal) [reply lib/experimental/class-gutenberg-rest-global-styles-controller.php]: I'll try it today and let you know!

**Comment by cbravobernal** [comment]
Thanks for your insights @ramonjd .
> For some reason it seems get_item_schema() from Core is being called first. I ran out of time to debug, but I'll carry on tomorrow.
I will debug it today.

> Oh, if this doesn't make it into WordPress 6.3, then we'd have to copy over any private methods to new 6-4 classes. Not a big deal, but a bit of a pain.

I will create a 6.4 compat folder and move everything there :-)

**Comment by cbravobernal** [comment]
@ramonjd I tested using register_rest_field, while it is working. We think that this code should be in Core in 6.4, so we would have to extend the classes instead of using this filters. I leave the code not to lose it:
```php
function gutenberg_register_behaviors_rest_field() {
    register_rest_field(
        'wp_global_styles',
        'behaviors',
        array(
            'get_callback'    => function() {
				$data  = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( 'custom' );
				if ( ! $data ) {
					$data = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( 'theme' );
				}
				$raw_data          = $data->get_raw_data();
				$behaviors = isset( $raw_data['behaviors'] ) ? $raw_data['behaviors'] : new stdClass();
				return $behaviors;
			},
			'update_callback' => function($data, $global_styles) {
				$raw_config = json_decode( $global_styles->post_content, true );
				$raw_config['behaviors'] = $data;
				return wp_update_post( array(
					'ID'           => $global_styles->ID,
					'post_content' => wp_json_encode( $raw_config ),
				) );

			},
            'schema'          => array(
				'description' => __( 'Global behaviors.', 'default' ),
				'type'        => array( 'object' ),
				'context'     => array( 'view', 'edit' ),
			),
        )
    );
}
add_action('rest_api_init', 'gutenberg_register_behaviors_rest_field');
```

**Comment by ramonjd** [comment]
> Could I ask you to take a look at the unit test failures in this PR? @c4rl0sbr4v0 and I have been trying to figure it out for a couple hours now but no dice. I'm asking you because you worked on Global Style Revisions and this is a related unit test.

So it looks like `WP_REST_Global_Styles_Revisions_Controller->get_item_schema()` (the Core version) is called first for the OPTIONS call. 

I tested this by adding the changes to `get_item_schema()` in this PR to Core. All tests passed, of course! 😄 

I'm not sure why yet.

I tried importing all the methods that call `$this->get_item_schema()`, hoping that it would force our class to use the local version, but haven't been able to crack it yet. Maybe there's a runtime order due to  `rest_api_init`.

cc @TimothyBJacobs if he has any insider tips.

The other option for now is to leave out the changes to that test. I was thinking of removing all of those tests now that they are running in Core anyway. 

But you could still test the response in `test_get_items()` - I tried that and it did work 😄

**Comment by cbravobernal** [comment]
> The other option for now is to leave out the changes to that test. I was thinking of removing all of those tests now anyway that they are running in Core anyway.

The issue is that if I leave the tests to the previous value, 9, it also fails 😓 

Removing tests that are now in Core, seems fine 😆

File: lib/load.php [hunk 6]
```diff
@@ -103,7 +107,7 @@ function gutenberg_is_experiment_enabled( $name ) {
 require_once __DIR__ . '/compat/wordpress-6.3/kses.php';
 
 // Experimental features.
-require __DIR__ . '/experimental/behaviors.php';
+remove_action( 'plugins_loaded', '_wp_theme_json_webfonts_handler' ); // Turns off WP 6.0's stopgap handler for Webfonts API.
```
Comment (michalczaplinski) [inline lib/load.php]: Why are we adding this? Seems unrelated to behaviors unless I'm missing something? 🤔

File: lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php [hunk 7]
```diff
@@ -21,7 +21,7 @@ class Gutenberg_REST_Global_Styles_Revisions_Controller_6_3 extends WP_REST_Cont
 	 * @since 6.3.0
 	 * @var string
 	 */
-	private $parent_post_type;
+	protected $parent_post_type;
```
Comment (michalczaplinski) [inline lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php:24]: Why do we need this change?

**Review by michalczaplinski — COMMENTED** [review]
Just a couple of small comments but otherwise this is ready 👍 

Awesome work, I'll push out a new PR today adding a few e2e tests for this.

File: lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php (hunk 7 above)
Comment (cbravobernal) [reply lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php:24]: Cause we use that variable to create the schema in `Gutenberg_REST_Global_Styles_Revisions_Controller_6_4` , that extends that class. At the beginning, that was only accessed by 6_3, so being private was fine, but, in order to access from a Class which is extending, we need it to be protected.

File: lib/load.php (hunk 6 above)
Comment (cbravobernal) [reply lib/load.php]: It is indeed unrelated 🤔 . We should remove it, as it seems that we took that from a rebase from trunk.

File: lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php (hunk 7 above)
Comment (michalczaplinski) [reply lib/compat/wordpress-6.3/class-gutenberg-rest-global-styles-revisions-controller-6-3.php:24]: Gotcha, yes that makes sense of course 🙂

File: lib/load.php (hunk 6 above)
Comment (michalczaplinski) [reply lib/load.php]: ok, I ll remove it!

Comment (michalczaplinski) [reply lib/load.php]: done [d9048d2](https://github.com/WordPress/gutenberg/pull/52370/commits/d9048d253482a50b7ae8b1cf8cd66f418ced5015)

**Review by michalczaplinski — APPROVED** [review]

**Comment by ramonjd** [comment]
Hello friends!

I got the test running over in https://github.com/WordPress/gutenberg/pull/52819

File: packages/edit-site/src/components/global-styles/screen-block.js [hunk 8]
```diff
@@ -61,12 +61,15 @@ const {
...
 	useHasBorderPanel,
+	__experimentalUseHasBehaviorsPanel: useHasBehaviorsPanel,
 	useGlobalSetting,
 	useSettingsForBlockElement,
 	useHasColorPanel,
 	useHasEffectsPanel,
 	useHasFiltersPanel,
 	useGlobalStyle,
+	__experimentalUseGlobalBehaviors: useGlobalBehaviors,
+	__experimentalBehaviorsPanel: StylesBehaviorsPanel,
```
Comment (youknowriad) [inline packages/edit-site/src/components/global-styles/screen-block.js:72]: Why are these prefixed with `__experimental` knowing that they are already private APIs

File: packages/edit-site/src/components/global-styles/screen-block.js [hunk 9]
```diff
@@ -91,10 +94,14 @@ function ScreenBlock( { name, variation } ) {
...
 	const [ rawSettings, setSettings ] = useGlobalSetting( '', name );
 	const settings = useSettingsForBlockElement( rawSettings, name );
+	const { inheritedBehaviors, setBehavior } = useGlobalBehaviors( name );
+	const { behavior } = useGlobalBehaviors( name, 'user' );
+
 	const blockType = getBlockType( name );
 	const blockVariations = useBlockVariations( name );
 	const hasTypographyPanel = useHasTypographyPanel( settings );
 	const hasColorPanel = useHasColorPanel( settings );
+	const hasBehaviorsPanel = useHasBehaviorsPanel( rawSettings, name );
```
Comment (youknowriad) [inline packages/edit-site/src/components/global-styles/screen-block.js:104]: Why is this using rawSettings instead of settings (like the other hooks).

File: packages/edit-site/src/components/global-styles/screen-block.js [hunk 10]
```diff
@@ -267,6 +274,14 @@ function ScreenBlock( { name, variation } ) {
 						onChange={ setStyle }
 						inheritedValue={ inheritedStyle }
 					/>
+					{ hasBehaviorsPanel && (
+						<StylesBehaviorsPanel
+							value={ behavior }
+							onChange={ setBehavior }
+							behaviors={ inheritedBehaviors }
+							blockName={ name }
+						></StylesBehaviorsPanel>
```
Comment (youknowriad) [inline packages/edit-site/src/components/global-styles/screen-block.js:283]: Nit: The element could have been autoclosed.

**Review by youknowriad — COMMENTED** [review]
I know this is old, but I just stumbled upon this code, so I wanted to leave some comments for a potential follow-up.
